### PR TITLE
nautilus: rgw: Use correct bucket info when put or get large object with swift.

### PR DESCRIPTION
**Pull request by xijiacun**
backport tracker: https://tracker.ceph.com/issues/50836

backport of #40296 

parent tracker: https://tracker.ceph.com/issues/49791

--- 

Fixes: https://tracker.ceph.com/issues/49791

Signed-off-by: zhiming zhang <zhangzhm1@chinatelecom.cn>
Signed-off-by: yupeng chen <chenyupeng@chinatelecom.cn>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by bel5f5** [comment]
@yehudasa Please help  to review the code.

**Comment by tchaikov** [comment]
@bel5f5 does this issue exist in master or any of the LTS branches? if yes, probably we should fix it in master first. and backport it to nautilus (and other LTS branches)?

**Comment by xijiacun** [comment]
> @xijiacun @bel5f5 does this issue exist in master or any of the LTS branches? if yes, probably we should fix it in master first. and backport it to nautilus (and other LTS branches)?

This issue does not exist in master, but exist in N and O branches.

**Comment by xijiacun** [comment]
> @xijiacun @bel5f5 does this issue exist in master or any of the LTS branches? if yes, probably we should fix it in master first. and backport it to nautilus (and other LTS branches)?

@tchaikov

**Comment by xijiacun** [comment]
> > @xijiacun @bel5f5 does this issue exist in master or any of the LTS branches? if yes, probably we should fix it in master first. and backport it to nautilus (and other LTS branches)?
> 
> This issue does not exist in master, but exist in N and O branches.

@tchaikov

File: src/rgw/rgw_rest_swift.cc [hunk 1]
```diff
@@ -858,8 +859,10 @@ int RGWPutObj_ObjStore_SWIFT::update_slo_segment_size(rgw_slo_entry& entry) {
       return r;
     }
     bucket = bucket_info.bucket;
+    pbucket_info = &bucket_info;
```
Comment (cbodley) [inline src/rgw/rgw_rest_swift.cc:862]: this is storing a pointer to `RGWBucketInfo bucket_info;` which goes out of scope on the next line

Comment (xijiacun) [reply src/rgw/rgw_rest_swift.cc:862]: OK, I fixed it in pr to octopus, https://github.com/ceph/ceph/pull/40296,

**Comment by smithfarm** [comment]
@xijiacun The standard procedure [1] in this case would be:

1. open fix targeting octopus - you already did that here: https://github.com/ceph/ceph/pull/40296
2. wait for the octopus fix to be reviewed, tested and merged
3. cherry-pick the octopus fix to nautilus using `git cherry-pick -x`

Would that be OK for you?

[1] https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst

**Comment by tchaikov** [comment]
@xijiacun sorry for the latency. i am not qualified for reviewing RGW changes like your fix.  so i'd defer the review to developers who have better understanding of RGW.

@smithfarm i am removing the DNM label, since the octopus fix has been merged.

**Review by smithfarm — CHANGES_REQUESTED** [review]
@xijiacun When I tried cherry-picking b6bf107e40de6179ff17076b51ac60332f210fae to nautilus, I got a conflict. Did you also?

The reason I ask is I noticed the commit message in this PR does not have a "Conflicts" section. For more information, see https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst#cherry-picking-rules where it says:

* the commit message generated by `git cherry-pick -x` must not be modified, except to add a "Conflicts" section below the "cherry picked from commit ..." line added by git
* the "Conflicts" section must mention all files where changes had to be made manually (not just conflicts flagged by git)
* the "Conflicts" section should also describe the manual changes that were made

**Comment by xijiacun** [comment]
> @xijiacun When I tried cherry-picking [b6bf107](https://github.com/ceph/ceph/commit/b6bf107e40de6179ff17076b51ac60332f210fae) to nautilus, I got a conflict. Did you also?
> 
> The reason I ask is I noticed the commit message in this PR does not have a "Conflicts" section, which ordinarily means the cherry-pick applied cleanly (?).
> 
> When you run `git cherry-pick -x` and need to resolve conflicts, git generates this "Conflicts" section when it generates the commit message. Unfortunately, as of a certain git version it also started commenting-out the Conflicts section that it generates, so now it's easy to forget to uncomment it.
> 
> For more information, see https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst#cherry-picking-rules where it says:
> 
> * the commit message generated by `git cherry-pick -x` must not be modified, except to add a "Conflicts" section below the "cherry picked from commit ..." line added by git
> * the "Conflicts" section must mention all files where changes had to be made manually (not just conflicts flagged by git)
> * the "Conflicts" section should also describe the manual changes that were made
> 
> and https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst#conflict-resolution which presents an example illustrating what a Conflicts section looks like.

yes, there is a conflict. I cherry-pick again to add conflicts section. @smithfarm

**Pull request merged**